### PR TITLE
improved IR

### PR DESCRIPTION
**Pull request by JeffBezanson**
#15181 rebased.

### CODE REVIEW

**Comment by vtjnash** [comment]
you need to not add [ci skip] to the commit message to so that CI can be allowed to run on this

**Comment by JeffBezanson** [comment]
Yes, that's kind of a remnant. I'm going to add a bit more to this before merging anyway.

File: base/stacktraces.jl [hunk 1]
```diff
@@ -166,7 +166,7 @@ function show_spec_linfo(io::IO, frame::StackFrame)
     else
         linfo = get(frame.outer_linfo)
         if isdefined(linfo, 8)
```
Comment (tkelman) [inline base/stacktraces.jl]: `isdefined(linfo, :specTypes)` ?

Comment (JeffBezanson) [inline base/stacktraces.jl]: :+1:

**Comment by KristofferC** [comment]
I have no idea what this is but I am excited anyway! :tada:

File: src/alloc.c [hunk 2]
```diff
@@ -272,6 +271,69 @@ JL_DLLEXPORT jl_value_t *jl_new_struct_uninit(jl_datatype_t *type)
...
+    jl_array_t *args = jl_lam_args((jl_expr_t*)ast);
+    size_t nslots = jl_array_len(vis);
+    size_t narg = jl_array_len(args);
+    li->nargs = narg;
+    li->isva = narg>0 && jl_is_rest_arg(jl_cellref(args, narg-1));
+    jl_value_t *gensym_types = jl_lam_gensyms((jl_expr_t*)ast);
+    size_t ngensym = (jl_is_array(gensym_types) ? jl_array_len(gensym_types) : jl_unbox_long(gensym_types));
+    li->slotnames = jl_alloc_cell_1d(nslots);
+    li->slottypes = jl_nothing;
+    li->slotflags = jl_alloc_array_1d(jl_array_uint8_type, nslots);
```
Comment (yuyichao) [inline src/alloc.c]: Missing wbs? (here and above)

File: src/alloc.c [hunk 3]
```diff
@@ -272,6 +271,69 @@ JL_DLLEXPORT jl_value_t *jl_new_struct_uninit(jl_datatype_t *type)
...
+    li->gensymtypes = jl_box_long(ngensym);
```
Comment (yuyichao) [inline src/alloc.c]: Here too

File: src/ast.c [hunk 4]
```diff
@@ -439,11 +441,11 @@ static jl_value_t *scm_to_julia_(fl_context_t *fl_ctx, value_t e, int eo)
...
             hd = car_(e);
-            if (hd == jl_ast_ctx(fl_ctx)->jlgensym_sym) {
-                size_t genid = numval(car_(cdr_(e)));
-                return jl_box_gensym(genid);
-            }
-            if (hd == jl_ast_ctx(fl_ctx)->null_sym && llength(e) == 1)
+            if (hd == jl_ast_ctx(fl_ctx)->jlgensym_sym)
+                return jl_box_gensym(numval(car_(cdr_(e))));
+            else if (hd == jl_ast_ctx(fl_ctx)->slot_sym)
+                return jl_new_struct(jl_slot_type, jl_box_long(numval(car_(cdr_(e)))), jl_any_type);
```
Comment (yuyichao) [inline src/ast.c]: the `jl_box_long` result can be free'd during `jl_new_struct`?

**Comment by JeffBezanson** [comment]
I should summarize:

- The compiler is faster; sysimg build should be noticeably faster.
- Code reflection is easier since more things are in named struct members instead of squirrelled away in expressions
- Compiler code is simpler
- When handling code, you don't need to go through a symbol lookup process. Everything in the IR tells you immediately what it is (i.e. local, global, or static parameter).
- Makes it easier to implement an even more efficient code representation in the future

File: src/builtins.c [hunk 5]
```diff
@@ -1084,11 +1078,11 @@ jl_value_t *jl_mk_builtin_func(const char *name, jl_fptr_t fptr)
 {
     jl_sym_t *sname = jl_symbol(name);
     jl_value_t *f = jl_new_generic_function_with_supertype(sname, jl_core_module, jl_builtin_type, 0);
-    jl_lambda_info_t *li = jl_new_lambda_info(jl_nothing, jl_emptysvec, jl_emptysvec, jl_core_module);
+    jl_lambda_info_t *li = jl_new_lambda_info(NULL, jl_emptysvec, jl_emptysvec, jl_core_module);
     li->fptr = fptr;
     li->name = sname;
     // TODO jb/functions: what should li->ast be?
-    li->ast = (jl_value_t*)jl_exprn(lambda_sym,0); jl_gc_wb(li, li->ast);
+    li->code = (jl_array_t*)jl_an_empty_cell; jl_gc_wb(li, li->code);
```
Comment (yuyichao) [inline src/builtins.c]: This wb can actually be removed but is fine to keep..

**Comment by JeffBezanson** [comment]
@yuyichao Thanks I will fix those.

**Pull request merged**